src/rp2/abstract_accounting_method.py

```diff
@@ -59,15 +76,22 @@ def set_partial_amount(self, acquired_lot: InTransaction, amount: RP2Decimal) ->
     def clear_partial_amount(self, acquired_lot: InTransaction) -> None:
         self.set_partial_amount(acquired_lot, ZERO)
 
-    def __iter__(self) -> "AccountingMethodIterator":
-        return AccountingMethodIterator(self.__acquired_lot_list, self.__up_to_index, self.__accounting_method.lot_candidates_order())
+    def add_selected_lot_to_heap(self, lot: InTransaction) -> None:
+        heap_item = (self.__accounting_method.heap_key(lot), lot)
+        heappush(self.__acquired_lot_heap, heap_item)
+
+    def __iter__(self) -> Union["AccountingMethodIterator", "HeapAccountingMethodIterator"]:
+        if self.__accounting_method.use_heap():  # pylint: disable=no-else-return
```

Comment:
Instead of conditional code can we capture this through inheritance? Specifically: 
* add a `_get_accounting_method_iterator` abstract method here;
* implement `_get_accounting_method_iterator` in FIFO, LIFO and HIFO plugins;
* call `_get_accounting_method_iterator` inside `__iter__` here.
* rename `AccountingMethodIterator` to `ListAccountingMethodIterator`
* add a `AbstractAccountingMethodIterator` superclass to `ListAccountingMethodIterator` and `HeapAccountingMethodIterator`
* have `__iter__` return `AbstractAccountingMethodIterator` rather than the `Union`

Reply:
Thanks for the feedback, I'll take a stab over the weekend.

Reply:
Updated with your suggestions - I had to add some accessors but otherwise should match what you have here.

Some tests are failing but I'm guessing that's a matter of fixing `heap_key()`/etc such that the sort is stable.

Reply:
Sounds good, thanks. After I merged #115 some conflicts appeared here. After they are fixed I'll take another look.

Reply:
Conflicts are resolved, needed a force push to update though

Reply:
OK, no worries.